---
title: "For Pub - ms_peptaibol"
author: "Katie Kyle"
date: "10/6/2021"
output: html_document
---

First run the my slightly modified (unpaired Trimmomatic, ITSx) DADA2 ITS pipeline from bash: 
  `"./unpairedITS_DADA2.R > unpairedITS_DADA2.R.out 2>&1"`

Adjust ITS2 fasta for manipulation: 
  `cat ITSx_psASV.ITS2.fasta | paste - - | sed -E 's/>ASV(.*)\|.\|/ASV\1\t/' > ITSx_psASV.ITS2.tsv`
  
Check if any ASVs did not pass ITSx and filter those out of phyloseq object:
  `cat ITSx_psASV_no_detections.fasta`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require("knitr")
#opts_knit$set(root.dir = "/Users/kkyle/Desktop/mountpoint/kekws/sda/data/ms_ITSartefact_analysis/trimmomatic/trimm.sw5.20/maits")
opts_knit$set(root.dir = "~/Desktop/mountpoint/klassenlab1/ms_peptaibol/TSinf/dada2")
# setwd("~/Desktop/mountpoint/klassenlab1/test/")

library("devtools")
#devtools::install_github("benjjneb/dada2", ref="v1.16") # change the ref argument to get other versions
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("phyloseq")
library(dada2)
library(phyloseq)
packageVersion("phyloseq")
library(ShortRead)
library(Biostrings)
library(dplyr)
library(tidyr)
library("data.table")
library(ggplot2)
# install.packages("remotes")
# remotes::install_github("vmikk/metagMisc")
library(metagMisc)
library(ggtext)
library(stringr)
#install.packages("decontam")
# BiocManager::install("decontam")
library(decontam); packageVersion("decontam")

set.seed(14)
```

## Time-series Infection ITS


```{r load ps.RDS}
getwd()
# load phyloseq object generated in unpairedITS_dada2.R
ps = readRDS("ps.RDS")
asvs.its2=read.csv("ITSx_psASV.ITS2.tsv",sep = "\t",header = F)

# Check if any ASVs did not pass ITSx and filter those out of phyloseq object:
setdiff(taxa_names(ps),asvs.its2$V1)
ps=prune_taxa(asvs.its2$V1,ps)
```

```{r ITSx processing}
# duplicated() determines which elements of a vector or data frame are duplicates of elements with smaller subscripts
w.dup=which(duplicated(asvs.its2$V3))   # these indices are duplicates of smaller indices
dup.m=match(asvs.its2[w.dup,]$V3,asvs.its2$V3)  # these are the indices that the dups match to
cbind(w.dup,dup.m)
asvs.its2[w.dup,]$V3 == asvs.its2[dup.m,]$V3

merge.ps=ps
# merged taxa/otu tables
for(i in seq(w.dup)){
  print(taxa_names(ps)[c(dup.m[i],w.dup[i])])           # match results must come first so duplicates are merged 
                                                        # into the smaller indexed ASV
  merge.ps=merge_taxa(merge.ps,taxa_names(ps)[c(w.dup[i],dup.m[i])])
}
```

```{r some checks1, echo=FALSE}
# otu_table(ps)[,c(3,4,7,26,12,9,23,24)]
```

```{r some checks2, echo=FALSE}
# #for(i in 1:nrow(otu_table(ps))) {
# for(i in 36) {
#   print("3 = 4 = 7 = 26")
#   print(otu_table(ps)[i,c(3,4,7,26)])
#   print(rowSums(otu_table(ps)[i,c(3,4,7,26)]))
#   
#   print(otu_table(merge.ps)[i,"ASV3"])
#   
#   print("9 = 12")
#   print(otu_table(ps)[i,c(9,12)])
#   print(rowSums(otu_table(ps)[i,c(9,12)]))
#   
#   print(otu_table(merge.ps)[i,"ASV9"])
#   
#   print("23 = 24")
#   print(otu_table(ps)[i,c(23,24)])
#   print(rowSums(otu_table(ps)[i,c(23,24)]))
#   
#   print(otu_table(merge.ps)[i,"ASV23"] )
# }

```

```{r}
otu.merge=otu_table(merge.ps)
# remove duplicate ASVs from ITSx refseq table
seqs.itsx = Biostrings::DNAStringSet(asvs.its2[-w.dup,]$V3)
names(seqs.itsx) = taxa_names(otu.merge)

# recall taxonomy for ITSx treated ASVs
x1 = otu.merge
colnames(x1) = as.character(seqs.itsx)
unite.ref="../../sh_general_release_dynamic_s_04.02.2020.fasta"
x1.tax=assignTaxonomy(x1,unite.ref,multithread = T, tryRC = T)
rownames(x1.tax) = colnames(otu.merge)

# remake phyloseq object with new ITSx seqs and merged ASV3/5
ps.itsx <- phyloseq(tax_table(x1.tax), otu_table(otu.merge, taxa_are_rows = F))
ps.itsx <- merge_phyloseq(ps.itsx,seqs.itsx)
taxa_names(ps.itsx) = taxa_names(otu.merge)

ps.itsx
```


```{r rename "cultivar"}
# set genus name to "Cultivar Fungus" if unclassified f__Agaricaceae
w.Fagar=which(tax_table(ps.itsx)[,"Family"] == "f__Agaricaceae")
w.gna=which(is.na(tax_table(ps.itsx)[,"Genus"]))
w.int=intersect(w.Fagar,w.gna)

tax_table(ps.itsx)[w.int,"Genus"]="Cultivar Fungus"
```

All before this point could be added to the automated unpairedITS_DADA2.R script if I can figure out how to convert the ITSx produced ITS2 fasta into the asvs.its2 TSV from within R. 


Not sure if rarefy should come before or after DECONTAM... Emily says after!


```{r DECONTAM}
metadat=read.csv("../TSinf.itsx.ps.sampledata.csv",header = T,row.names = 1)
sample_data(ps.itsx)=metadat
sample_data(ps.itsx)

w.quant0=which(sample_data(ps.itsx)$quant_reading == 0)
sample_data(ps.itsx)$quant_reading[w.quant0] = 0.00001
contamdf.freq=isContaminant(ps.itsx, method="frequency", conc = "quant_reading")
table(contamdf.freq$contaminant)  # no contaminants this way

contamdf.prev <- isContaminant(ps.itsx, method="prevalence", neg="is.neg")
table(contamdf.prev$contaminant)  # also none

contamdf.both.batch=isContaminant(ps.itsx, method="combined", batch="batch", conc="quant_reading", neg="is.neg")
table(contamdf.both.batch$contaminant)
# contamdf.both.batch    # also also none

contamdf.both=isContaminant(ps.itsx, method="combined", conc="quant_reading", neg="is.neg")
table(contamdf.both$contaminant)
# contamdf.both    # also also also none

contamdf.min.batch=isContaminant(ps.itsx, method="minimum", batch="batch", conc="quant_reading", neg="is.neg")
table(contamdf.min.batch$contaminant)
# contamdf.min.batch    # also also also also none

contamdf.eith.batch=isContaminant(ps.itsx, method="either", batch="batch", conc="quant_reading", neg="is.neg")
table(contamdf.eith.batch$contaminant)
# contamdf.min.batch    # also also also also also none

contamdf.eith=isContaminant(ps.itsx, method="either", conc="quant_reading", neg="is.neg")
table(contamdf.eith$contaminant)
# contamdf.min.batch    # also also also also also also none

# w=which(contamdf.prev$contaminant)
# contamdf.prev[w,]
# contam19seq=rownames(contamdf.prev)[19]
# 
# which(rownames(tax_table(ps)) == contam19seq)
# tax_table(ps)[w,]   # ok good! the only contams are that suspicious all G seq and then an unclass. fungi

## ----prune contaminants -------
# ps
# ps.noncontam <- prune_taxa(!contamdf.prev$contaminant, ps)
# ps.noncontam
```
No contaminants so we can remove the negative control samples. 
```{r}
tf.blank=sample_data(ps.itsx)$sample_type == "blank"
sample_data(ps.itsx)[tf.blank,]
ps.itsx.no_nc=prune_samples(!tf.blank, ps.itsx)
```



```{r rarefy}
# ##----- rarefy -----
track=read.csv("track.csv",row.names = 1,header=T)
sort(sample_sums(ps.itsx.no_nc))
barplot(sort(sample_sums(ps.itsx.no_nc))); abline(h=4099, col='red')
 #       # rarefy to 4099 reads
set.seed(14); ps.itsx.no_nc.raref = rarefy_even_depth(ps.itsx.no_nc, sample.size = 4099)
#                   2 samples removedbecause they contained fewer reads than `sample.size`.
#                   Up to first five removed samples are: 
#                       TSe26TSe04
sort(sample_sums(ps.itsx.no_nc.raref))

```

```{r filter highly unclass ASVs}
which(tax_table(ps.itsx.no_nc.raref)[,"Kingdom"] != "k__Fungi")   # all classified as fungi
w.pna=which(is.na(tax_table(ps.itsx.no_nc.raref))[,"Phylum"])     # two ASVs unclassified except for k__Fungi (29, 35)
tax_table(ps.itsx.no_nc.raref)[w.pna,]
ps.itsx.no_nc.raref.filt=subset_taxa(ps.itsx.no_nc.raref, !is.na(Phylum)) # now 31 taxa
```

```{r threshold filtering i.e. ASVs <1% abundance in at least one sample}
library(metagMisc)
ps.itsx.no_nc.raref.filt.gt1perc=phyloseq_filter_sample_wise_abund_trim(ps.itsx.raref.filt, minabund = 41, relabund = FALSE, rm_zero_OTUs = TRUE)  # 40.99 reads is 1% of 4099 (which is what smaples are raref to)
tax_table(ps.itsx.no_nc.raref.filt.gt1perc) # 14 ASVs now
```
```{r}
# tax_table(ps.itsx.raref.filt)[,"Genus"]
# tax_table(ps.itsx.raref.filt.gt1perc)[,"Genus"]
```

1. adjust unclassified genera
2. combine cultivar
3. glom by genus
4. filter based on gt1perc genera (bc glommed, prev filter was by ASV?)
        - dont think this is really necessary... should be fine to just plot since already filtered on 1% thresh
5. calc and check prevalence
6. plot

```{r 1. adjust unclassified genera}
# ## 1. adjust unclassified genera
# # define unclassified genera by family so no taxa lost
gen.na=is.na(tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[,"Genus"])
if(length(which(gen.na)) > 0) {
  unclass.string=paste("unclassified",tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[gen.na,"Family"])
  tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[gen.na,"Genus"] = unclass.string
}
```

```{r 2. combine cultivar}
# ## 2. combine cultivar
# # combine leucocoprinus and unclassified agaricaecae
unique(tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[,"Genus"])
w.cult=which(tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[,"Genus"] == "g__Leucocoprinus" |
               tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[,"Genus"] == "g__Leucoagaricus" |
               tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[,"Genus"] == "unclassified f__Agaricaceae") 
if(length(w.cult) > 0) {
  tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[w.cult,"Genus"] = "Cultivar Fungus"
  tax_table(ps.itsx.no_nc.raref.filt.gt1perc)[w.cult,] 
}
```

```{r 3. glom by genus}
# ## 3. glom by genus
ps.itsx.no_nc.raref.filt.gt1perc.gglom=tax_glom(ps.itsx.no_nc.raref.filt.gt1perc, "Genus" , NArm=TRUE)
ps.itsx.no_nc.raref.filt.gt1perc.gglom     # now 5 taxa
tax_table(ps.itsx.no_nc.raref.filt.gt1perc.gglom)[,"Genus"]
```

```{r 4. filter based on gt1perc genera}
# ## 4. filter based on gt1perc genera
# gt1perc.asvs=paste0("ASV",c(1,6,5,462,16,479,20,476,472,460))
# filtNot0out.glom.genus = prune_taxa(gt1perc.asvs, filtNot0out.glom.genus)
# tax_table(filtNot0out.glom.genus)
```

```{r 5. calc and check prevalence}
# # prevalence 
prev.itsx.no_nc.raref.filt.gt1perc.gglom = apply(X = otu_table(ps.itsx.no_nc.raref.filt.gt1perc.gglom),
                # MARGIN = 2,   # this needs to be 2 if taxa are columns instead of rows!
                 MARGIN = ifelse(taxa_are_rows(ps.itsx.no_nc.raref.filt.gt1perc.gglom), 
                                 yes = 1, no = 2),# FANCY https://www.bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/doc/MicrobiomeWorkshopII.html#prevalence-filtering
                 FUN = function(x){sum(x > 0)})
# # Add taxonomy and total read counts to this data.frame
prev.itsx.no_nc.raref.filt.gt1perc.gglom = data.frame(Prevalence = prev.itsx.no_nc.raref.filt.gt1perc.gglom,
                      TotalAbundance = taxa_sums(ps.itsx.no_nc.raref.filt.gt1perc.gglom),
                      # RelGlomAbund.perc = 100*taxa_sums(ps.itsx.raref.filt.gt1perc.gglom)/sum( taxa_sums( ps.itsx.raref.filt.gt1perc.gglom )),
                      # RelNonCultAbund.perc=100*taxa_sums(filtNot0out.glom.genus)/sum(taxa_sums(nonCultivar.raref.itsx.ps)),
                      RelRarefAbund.perc=100*taxa_sums(ps.itsx.no_nc.raref.filt.gt1perc.gglom)/sum(taxa_sums(ps.itsx.no_nc.raref.filt.gt1perc.gglom)),
                      tax_table(ps.itsx.no_nc.raref.filt.gt1perc.gglom))
prev.itsx.no_nc.raref.filt.gt1perc.gglom[order(-prev.itsx.no_nc.raref.filt.gt1perc.gglom$Prevalence),]
prev.itsx.no_nc.raref.filt.gt1perc.gglom[order(-prev.itsx.no_nc.raref.filt.gt1perc.gglom$RelRarefAbund.perc),]
```

```{r plot}
plot_bar(ps.itsx.no_nc.raref.filt.gt1perc.gglom, fill = "Genus") +
  facet_wrap(~treatment,scales = "free_x")
```

```{r look at trich only, fig.width=12, fig.height=6}
# # ##----- investigate trichoderma only---- 
# # fam.hypoc.w=which(taxtab.print[,"Family"] == "f__Hypocreaceae")
# # taxtab.print[fam.hypoc.w,]  # all f__Hypocreaceae are Trichoderma
# # 
# # trichonly.itsx.ps.filt.raref=subset_taxa(itsx.ps.filt.raref, Genus=="g__Trichoderma")
# # trichonly.itsx.ps.filt.raref.prop=subset_taxa(itsx.ps.filt.raref.prop, Genus=="g__Trichoderma")
# # 
# # trichonly.itsx.ps.filt.raref
# # sort(sample_sums(trichonly.itsx.ps.filt.raref)) 
# # sum(sample_sums(trichonly.itsx.ps.filt.raref))
# # 
# # 
# # plot_bar(trichonly.itsx.ps.filt.raref, fill = "Genus") + 
# #   theme(axis.text=element_text(size=14),legend.text = element_text(size=9)) +
# #   facet_wrap(~treatment, scales="free_x") +
# #   ggtitle("Time Series Trich. Infection - Trich only ITSx, filtered, rarefy 3981 -- Absolute Abund.") +
# #   scale_fill_manual(values=cbgreen)
# # 
# # plot_bar(trichonly.itsx.ps.filt.raref.prop, fill = "Genus") + 
# #   theme(axis.text=element_text(size=14),legend.text = element_text(size=9)) +
# #   facet_wrap(~treatment, scales="free_x") +
# #   ggtitle("Time Series Trich. Infection - Trich only ITSx, filtered, rarefy 3981 -- Rel. Abund.") +
# #   scale_fill_manual(values=cbgreen)
```

**Need to make an "Other" category so stacked bar charts sum to 1.0. **